crates/serialize_hierarchy/src/lib.rs

```diff
@@ -1,183 +1,302 @@
+    {
+        Err(Error::TypeDoesNotSupportSerialization {
+            type_name: "Vec",
+            path: path.to_string(),
+        })
     }
 
-    fn deserialize_hierarchy(&mut self, field_path: &str, _data: SerializedValue) -> Result<()> {
-        bail!("cannot access Vec with path: {}", field_path)
+    fn deserialize_path<S>(
+        &mut self,
+        path: &str,
+        _data: S::Serialized,
+    ) -> Result<(), Error<S::Error>>
+    where
+        S: Serializer,
+        S::Error: error::Error,
+    {
+        Err(Error::TypeDoesNotSupportDeserialization {
+            type_name: "Vec",
+            path: path.to_string(),
+        })
     }
 
-    fn exists(_field_path: &str) -> bool {
+    fn exists(_path: &str) -> bool {
         false
     }
 
-    fn get_hierarchy() -> HierarchyType {
-        HierarchyType::GenericStruct
+    fn get_fields() -> BTreeSet<String> {
+        [String::new()].into()
     }
 }
 
 macro_rules! serialize_hierarchy_primary_impl {
     ($type:ty) => {
         impl SerializeHierarchy for $type {
-            fn serialize_hierarchy(
-                &self,
-                field_path: &str,
-                _format: Format,
-            ) -> Result<SerializedValue> {
-                bail!(
-                    "cannot access {} with path: {}",
-                    stringify!($type),
-                    field_path
-                )
+            fn serialize_path<S>(&self, path: &str) -> Result<S::Serialized, Error<S::Error>>
+            where
+                S: Serializer,
+                S::Error: error::Error,
+            {
+                Err(Error::TypeDoesNotSupportSerialization {
+                    type_name: stringify!($type),
+                    path: path.to_string(),
+                })
             }
 
-            fn deserialize_hierarchy(
+            fn deserialize_path<S>(
                 &mut self,
-                field_path: &str,
-                _data: SerializedValue,
-            ) -> Result<()> {
-                bail!(
-                    "cannot access {} with path: {}",
-                    stringify!($type),
-                    field_path
-                )
+                path: &str,
+                _data: S::Serialized,
+            ) -> Result<(), Error<S::Error>>
+            where
+                S: Serializer,
+                S::Error: error::Error,
+            {
+                Err(Error::TypeDoesNotSupportDeserialization {
+                    type_name: stringify!($type),
+                    path: path.to_string(),
+                })
             }
 
-            fn exists(_field_path: &str) -> bool {
+            fn exists(_path: &str) -> bool {
                 false
             }
 
-            fn get_hierarchy() -> HierarchyType {
-                HierarchyType::Primary {
-                    name: stringify!($type),
-                }
+            fn get_fields() -> BTreeSet<String> {
+                [String::new()].into()
```

Comment:
Why doesn't this return an empty set?
This introduces trailing periods into the paths.

Reply:
Done